File: web/gray-code.ts
export const STATE_COUNT = 126;

const BOX_COUNT = 9;
const OCCUPIED_COUNT = 5;
const INITIAL_MASK = (1 << OCCUPIED_COUNT) - 1;
const LOGICAL_TO_BOX = [4, 2, 7, 5, 3, 6, 8, 9, 1] as const;

export interface Movement {
	fromBox: number;
	toBox: number;
}

export interface StageState {
	index: number;
	stateNumber: number;
	stateCount: number;
	occupied: number[];
	movement: Movement | null;
	instruction: string;
}

function bitCount(value: number): number {
	let remaining = value;
	let count = 0;
	while (remaining !== 0) {
		count += remaining & 1;
		remaining >>>= 1;
	}
	return count;
}

function remapToBoxes(logicalMask: number): number {
	let mask = 0;
	for (let logicalIndex = 0; logicalIndex < LOGICAL_TO_BOX.length; logicalIndex += 1) {
		if ((logicalMask & (1 << logicalIndex)) !== 0) {
			mask |= 1 << (LOGICAL_TO_BOX[logicalIndex] - 1);
		}
	}
	return mask;
}

function generateStates(): readonly number[] {
	const states: number[] = [];
	for (let binary = 0; binary < 1 << BOX_COUNT; binary += 1) {
		const gray = binary ^ (binary >> 1);
		if (bitCount(gray) === OCCUPIED_COUNT) {
			states.push(remapToBoxes(gray));
		}
	}

	const start = states.indexOf(INITIAL_MASK);
	if (start < 0) {
		throw new Error("Gray-code sequence is missing its initial state");
	}

	const rotated = [...states.slice(start), ...states.slice(0, start)];
	const result = [rotated[0], ...rotated.slice(1).reverse()];
	validateStates(result);
	return Object.freeze(result);
}

function validateStates(states: readonly number[]): void {
	if (states.length !== STATE_COUNT || new Set(states).size !== STATE_COUNT) {
		throw new Error("Gray-code sequence must contain 126 unique states");
	}
	if (states[0] !== INITIAL_MASK) {
		throw new Error("Gray-code sequence must start with boxes 1-5 occupied");
	}
	for (let index = 1; index < states.length; index += 1) {
		if (bitCount(states[index - 1] ^ states[index]) !== 2) {
			throw new Error("Adjacent Gray-code states must exchange exactly one box");
		}
	}
	const firstMove = movementBetween(states[0], states[1]);
	if (firstMove.fromBox !== 4 || firstMove.toBox !== 7) {
		throw new Error("Gray-code sequence must begin with move 4 to 7");
	}
}

function boxesForMask(mask: number): number[] {
	const boxes: number[] = [];
	for (let box = 1; box <= BOX_COUNT; box += 1) {
		if ((mask & (1 << (box - 1))) !== 0) {
			boxes.push(box);
		}
	}
	return boxes;
}

export function movementBetween(beforeMask: number, afterMask: number): Movement {
	const changed = beforeMask ^ afterMask;
	if (bitCount(changed) !== 2) {
		throw new Error("A movement must exchange exactly one occupied box");
	}
	const source = changed & beforeMask;
	const target = changed & afterMask;
	return {
		fromBox: Math.log2(source) + 1,
		toBox: Math.log2(target) + 1,
	};
}

export const STAGE_STATES = generateStates();

export function createStageState(index: number, movement: Movement | null = null): StageState {
	if (!Number.isInteger(index) || index < 0 || index >= STATE_COUNT) {
		throw new RangeError(`State index must be between 0 and ${STATE_COUNT - 1}`);
	}

	const occupied = boxesForMask(STAGE_STATES[index]);
	const resolvedMovement =
		movement ?? (index > 0 ? movementBetween(STAGE_STATES[index - 1], STAGE_STATES[index]) : null);
	const instruction = resolvedMovement
		? `State ${index + 1}/${STATE_COUNT} · Occupied {${occupied.join(",")}} · Move box ${resolvedMovement.fromBox} to box ${resolvedMovement.toBox}`
		: `State 1/${STATE_COUNT} · Occupied {${occupied.join(",")}} · Initial setup`;

	return {
		index,
		stateNumber: index + 1,
		stateCount: STATE_COUNT,
		occupied,
		movement: resolvedMovement,
		instruction,
	};
}

File: src/gray_code.rs
pub const STATE_COUNT: usize = 126;

const BOX_COUNT: u32 = 9;
const OCCUPIED_COUNT: u32 = 5;
const INITIAL_MASK: u16 = (1_u16 << OCCUPIED_COUNT) - 1;

// The filtered binary-reflected Gray code is invariant under bit permutation.
// This permutation keeps the public box numbers unchanged while shortening the
// route on the Stage 8 layout. The sequence is rotated to start at boxes 1-5.
const LOGICAL_TO_BOX: [u8; BOX_COUNT as usize] = [4, 2, 7, 5, 3, 6, 8, 9, 1];

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct GrayState {
    occupied_mask: u16,
}

impl GrayState {
    pub const fn occupied_mask(self) -> u16 {
        self.occupied_mask
    }

    pub fn occupied_boxes(self) -> [u8; OCCUPIED_COUNT as usize] {
        let mut boxes = [0; OCCUPIED_COUNT as usize];
        let mut output = 0;
        for box_index in 0..BOX_COUNT {
            if self.occupied_mask & (1 << box_index) != 0 {
                boxes[output] = box_index as u8 + 1;
                output += 1;
            }
        }
        debug_assert_eq!(output, OCCUPIED_COUNT as usize);
        boxes
    }

    pub fn movement_to(self, next: Self) -> Movement {
        let changed = self.occupied_mask ^ next.occupied_mask;
        assert_eq!(
            changed.count_ones(),
            2,
            "adjacent Gray-code states must exchange exactly one box"
        );

        let source = changed & self.occupied_mask;
        let target = changed & next.occupied_mask;
        assert_eq!(source.count_ones(), 1);
        assert_eq!(target.count_ones(), 1);

        Movement {
            from_box: source.trailing_zeros() as u8 + 1,
            to_box: target.trailing_zeros() as u8 + 1,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Movement {
    pub from_box: u8,
    pub to_box: u8,
}

pub fn generate_states() -> Vec<GrayState> {
    let mut states = (0_u16..1_u16 << BOX_COUNT)
        .map(|binary| binary ^ (binary >> 1))
        .filter(|gray| gray.count_ones() == OCCUPIED_COUNT)
        .map(|gray| GrayState {
            occupied_mask: remap_to_boxes(gray),
        })
        .collect::<Vec<_>>();

    let start = states
        .iter()
        .position(|state| state.occupied_mask == INITIAL_MASK)
        .expect("the Gray code contains the initial 1-5 combination");
    states.rotate_left(start);

    // Keep 1-5 as the first state and take the shorter direction around the
    // cyclic code. This makes the first instruction 4 -> 7.
    states[1..].reverse();

    debug_assert_eq!(states.len(), STATE_COUNT);
    debug_assert!(
        states
            .windows(2)
            .all(|pair| (pair[0].occupied_mask ^ pair[1].occupied_mask).count_ones() == 2)
    );
    states
}

fn remap_to_boxes(logical_mask: u16) -> u16 {
    LOGICAL_TO_BOX
        .into_iter()
        .enumerate()
        .filter(|(logical_index, _)| logical_mask & (1 << logical_index) != 0)
        .fold(0, |mask, (_, box_number)| mask | (1 << (box_number - 1)))
}

fn format_boxes(boxes: [u8; OCCUPIED_COUNT as usize]) -> String {
    boxes
        .into_iter()
        .map(|box_number| box_number.to_string())
        .collect::<Vec<_>>()
        .join(",")
}

pub fn format_init(state: GrayState) -> String {
    format!(
        "State 1/{STATE_COUNT} · Occupied {{{}}} · Initial setup",
        format_boxes(state.occupied_boxes())
    )
}

pub fn format_move(state_number: usize, before: GrayState, after: GrayState) -> String {
    assert!((1..=STATE_COUNT).contains(&state_number));
    let movement = before.movement_to(after);
    format!(
        "State {state_number}/{STATE_COUNT} · Occupied {{{}}} · Move box {} to box {}",
        format_boxes(after.occupied_boxes()),
        movement.from_box,
        movement.to_box
    )
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::*;

    #[test]
    fn sequence_covers_every_five_of_nine_combination_once() {
        let states = generate_states();
        let masks = states
            .iter()
            .map(|state| state.occupied_mask())
            .collect::<HashSet<_>>();
        let expected = (0_u16..1_u16 << BOX_COUNT)
            .filter(|mask| mask.count_ones() == OCCUPIED_COUNT)
            .collect::<HashSet<_>>();

        assert_eq!(states.len(), STATE_COUNT);
        assert_eq!(masks, expected);
    }

    #[test]
    fn every_transition_moves_exactly_one_player() {
        let states = generate_states();
        for pair in states.windows(2) {
            let movement = pair[0].movement_to(pair[1]);
            assert_ne!(movement.from_box, movement.to_box);
        }
        assert_eq!(
            (states[STATE_COUNT - 1].occupied_mask() ^ states[0].occupied_mask()).count_ones(),
            2,
            "the generated Gray code is cyclic"
        );
    }

    #[test]
    fn sequence_starts_with_an_intuitive_single_move() {
        let states = generate_states();
        assert_eq!(states[0].occupied_boxes(), [1, 2, 3, 4, 5]);
        assert_eq!(
            states[0].movement_to(states[1]),
            Movement {
                from_box: 4,
                to_box: 7,
            }
        );
        assert_eq!(
            format_init(states[0]),
            "State 1/126 · Occupied {1,2,3,4,5} · Initial setup"
        );
        assert_eq!(
            format_move(2, states[0], states[1]),
            "State 2/126 · Occupied {1,2,3,5,7} · Move box 4 to box 7"
        );
    }
}
